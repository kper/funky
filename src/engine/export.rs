use wasm_parser::core::{ExternalKindType, ExportEntry};

#[derive(Debug, Clone, PartialEq)]
pub struct ExportInstance {
    pub name: String,
    pub value: ExternalKindType,
}

impl From<&ExportEntry> for ExportInstance {
    fn from(state: &ExportEntry) -> ExportInstance {
        ExportInstance {
            name: state.name.clone(),
            value: state.kind,
        }
    }
}
