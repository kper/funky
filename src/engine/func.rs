use wasm_parser::core::{FunctionBody, FunctionSignature};

#[derive(Debug, Clone)]
pub struct FuncInstance {
    //FIXME Add HostFunc
    pub ty: FunctionSignature,
    pub code: FunctionBody,
}
