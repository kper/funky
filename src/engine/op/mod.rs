mod local_get;
mod local_set;
mod local_tee;
mod global_get;
mod global_set;
mod select;
mod memory_size;
mod memory_grow;
mod block;
mod br;
mod call;
mod call_indirect;
