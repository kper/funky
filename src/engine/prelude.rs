pub use crate::convert;
pub use crate::debugger::ProgramState;
pub use crate::debugger::{ProgramCounter, RelativeProgramCounter};
pub use crate::store::Store;
pub use crate::engine::stack::CtrlStackContent::*;
pub use crate::engine::stack::{Label, Frame};
pub use crate::operations::*;
pub use crate::page::Page;
pub use crate::value::{Arity, Value, Value::*};
pub use crate::PAGE_SIZE;
pub use anyhow::{anyhow, Context, Result};
pub use std::fmt;
pub use wasm_parser::core::Instruction::*;
pub use wasm_parser::core::*;
pub use wasm_parser::Module;
