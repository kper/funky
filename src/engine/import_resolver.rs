use crate::engine::{GlobalInstance, TableInstance};
use anyhow::{anyhow, Result};
use std::collections::HashMap;
//use wasm_parser::core::ImportEntry;

pub type Imports = Vec<Import>;

type Module = String;
type Name = String;

//TODO add more types
#[derive(Debug)]
pub enum Import {
    Global(Module, Name, GlobalInstance),
    Table(Module, Name, TableInstance),
}

/// Private lookup table for
/// modules and names
#[derive(Debug, Default)]
struct LookupTable {
    modules: HashMap<(Module, Name), Import>,
}

impl LookupTable {
    pub fn lookup(&self, module: impl Into<String>, name: impl Into<String>) -> Option<&Import> {
        if let Some(m) = self.modules.get(&(module.into(), name.into())) {
            return Some(m);
        }

        None
    }
}

/// The `ImportResolver` matches the `wasm_parser::core::ImportDesc`
/// with the actual given imports.
#[derive(Debug)]
pub(crate) struct ImportResolver {
    imports: LookupTable,
}

impl ImportResolver {
    pub fn new() -> Self {
        Self {
            imports: LookupTable::default(),
        }
    }

    /// Get the imported global by module and name
    pub fn resolve_global(&self, module: &String, name: &String) -> Result<GlobalInstance> {
        debug!("resolve global {} {}", module, name);

        if let Some(Import::Global(_, _, instance)) = self.imports.lookup(module, name) {
            return Ok(instance.clone());
        }

        Err(anyhow!("Cannot find global for {} {}", module, name))
    }

    /// Get the imported table by module and name
    pub fn resolve_table(&self, module: &String, name: &String) -> Result<TableInstance> {
        debug!("resolve table {} {}", module, name);

        if let Some(Import::Table(_, _, instance)) = self.imports.lookup(module, name) {
            return Ok(instance.clone());
        }

        Err(anyhow!("Cannot find table for {} {}", module, name))
    }

    pub fn inject_global(
        &mut self,
        module: Module,
        name: Name,
        instance: &GlobalInstance,
    ) -> Result<()> {
        self.imports.modules.insert(
            (module.clone(), name.clone()),
            Import::Global(module, name, instance.clone()),
        );

        Ok(())
    }

    pub fn inject_table(
        &mut self,
        module: Module,
        name: Name,
        instance: &TableInstance,
    ) -> Result<()> {
        self.imports.modules.insert(
            (module.clone(), name.clone()),
            Import::Table(module, name, instance.clone()),
        );

        Ok(())
    }
}
