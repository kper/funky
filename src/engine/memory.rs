use crate::engine::Page;
use std::fmt;

#[derive(Clone)]
pub struct MemoryInstance {
    pub data: Vec<u8>,
    pub max: Option<u32>,
}

// Overwritten debug implementation
// Because `data` can have a lot of entries, which
// can be a problem when printing
impl fmt::Debug for MemoryInstance {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("MemoryInstance")
            .field("data (only length)", &self.data.len())
            .field("max", &self.max)
            .finish()
    }
}

/// Returns Err when paging failed
/// Returns new length in pages
/// https://webassembly.github.io/spec/core/exec/modules.html#growing-memories
pub(crate) fn grow_memory(instance: &mut MemoryInstance, n: Page) -> Result<Page, ()> {
    if n.is_zero() {
        return Ok(Page::from_count(instance.data.len()));
    }

    let len = n + Page::from_count(instance.data.len());

    if len.pages() > usize::pow(2, 16) {
        error!("Length exceeded. Too many memory pages");
        return Err(());
    }

    if let Some(max) = instance.max {
        debug!("Checking limit len {:?} < max {}", len.pages(), max);
        if len.pages() > max as usize {
            error!("Memory growing failed. Limit exceded");
            return Err(());
        }
    }

    let new_length = Page::from_count(instance.data.len()) + n;
    debug!("Resize by {} bytes", new_length.elements());

    // Create new vec and fill it with 0u8
    let extension = vec![0u8; n.elements()];

    // Append the new vec to the instance
    instance.data.extend_from_slice(&extension);

    // Return pages
    Ok(new_length)
}
