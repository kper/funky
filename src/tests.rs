use crate::engine::Value::*;
use crate::engine::*;
use insta::assert_snapshot;
use std::cell::RefCell;
use std::rc::Rc;
use validation::validate;
use wasm_parser::core::*;
use wasm_parser::{parse, read_wasm, Module};

macro_rules! test_file_engine {
    ($fs_name:expr) => {
        let file = read_wasm!(&format!("tests/{}", $fs_name));
        let module = parse(file).expect("Parsing failed");
        assert!(validate(&module).is_ok());

        let instance = ModuleInstance::new(&module);
        let engine = Engine::new(instance, &module);

        assert_snapshot!($fs_name, format!("{:#?}", engine));
    };
}

macro_rules! test_run_engine {
    ($fs_name:expr, $num_f:expr, $init:expr) => {{
        let file = read_wasm!(&format!("tests/{}", $fs_name));
        let module = parse(file).expect("Parsing failed");
        assert!(validate(&module).is_ok());

        let instance = ModuleInstance::new(&module);
        let mut engine = Engine::new(instance, &module);

        assert_snapshot!($fs_name, format!("{:#?}", engine));

        engine.invoke_exported_function($num_f, $init);
        engine
    }};
}

macro_rules! allocation {
    ($sections:expr) => {{
        let module = Module {
            sections: $sections,
        };

        let instance = ModuleInstance::new(&module);
        let engine = Engine::new(instance, &module);

        engine
    }};
}

#[test]
fn test_allocation() {
    allocation!(vec![]); //no panic
}

#[test]
fn test_allocation_funcs() {
    let sig = FunctionSignature {
        param_types: vec![ValueType::I32],
        return_types: vec![ValueType::I32],
    };

    let body = FunctionBody {
        locals: vec![],
        code: vec![Instruction::Ctrl(CtrlInstructions::OP_NOP)],
    };

    let engine = allocation!(vec![
        Section::Type(TypeSection {
            entries: vec![sig.clone()]
        }),
        Section::Function(FunctionSection { types: vec![0] }),
        Section::Code(CodeSection {
            entries: vec![body.clone()]
        })
    ]);

    // Module instance has an entry for type
    // Module instance has an entry for code
    // Module instance has an entry for funcaddrs

    let mi = engine.module.borrow();
    assert_eq!(1, mi.fn_types.len());
    assert_eq!(sig, mi.fn_types[0]);
    assert_eq!(1, mi.code.len());
    assert_eq!(body, mi.code[0]);
    assert_eq!(1, mi.funcaddrs.len());
    assert_eq!(Some(&0), mi.funcaddrs.get(0));

    // Store has an entry for func instance

    assert_eq!(1, engine.store.funcs.len());
    assert_eq!(sig, engine.store.funcs[0].ty);
    assert_eq!(body, engine.store.funcs[0].code);
}

#[test]
fn test_allocation_tables_zero() {
    let engine = allocation!(vec![Section::Table(TableSection {
        entries: vec![TableType {
            element_type: 0x70,
            limits: Limits::Zero(10)
        }]
    })]);

    // Module instance has an entry in tableaddrs with 0
    // Store has a table instance

    assert_eq!(1, engine.module.borrow().tableaddrs.len());
    assert_eq!(Some(&0), engine.module.borrow().tableaddrs.get(0));

    assert_eq!(1, engine.store.tables.len());
    assert_eq!(10, engine.store.tables[0].elem.len());
    assert!(engine.store.tables[0].elem.iter().all(|w| w == &None));
    assert_eq!(None, engine.store.tables[0].max);
}

#[test]
fn test_allocation_tables_one() {
    let engine = allocation!(vec![Section::Table(TableSection {
        entries: vec![TableType {
            element_type: 0x70,
            limits: Limits::One(10, 20)
        }]
    })]);

    // Module instance has an entry in tableaddrs with 0
    // Store has a table instance

    assert_eq!(1, engine.module.borrow().tableaddrs.len());
    assert_eq!(Some(&0), engine.module.borrow().tableaddrs.get(0));

    assert_eq!(1, engine.store.tables.len());
    assert_eq!(10, engine.store.tables[0].elem.len());
    assert!(engine.store.tables[0].elem.iter().all(|w| w == &None));
    assert_eq!(Some(20), engine.store.tables[0].max);
}

#[test]
fn test_allocation_memories_zero() {
    let engine = allocation!(vec![Section::Memory(MemorySection {
        entries: vec![MemoryType {
            limits: Limits::Zero(10)
        }]
    })]);

    // Module instance has an entry in memaddrs with 0
    // Store has a memory instance

    assert_eq!(1, engine.module.borrow().memaddrs.len());
    assert_eq!(Some(&0), engine.module.borrow().memaddrs.get(0));

    assert_eq!(1, engine.store.memory.len());
    assert_eq!(10 * 1024 * 64, engine.store.memory[0].data.len());
    assert!(engine.store.memory[0].data.iter().all(|w| w == &0u8));
    assert_eq!(None, engine.store.memory[0].max);
}

#[test]
fn test_allocation_memories_one() {
    let engine = allocation!(vec![Section::Memory(MemorySection {
        entries: vec![MemoryType {
            limits: Limits::One(10, 20)
        }]
    })]);

    // Module instance has an entry in memaddrs with 0
    // Store has a memory instance

    assert_eq!(1, engine.module.borrow().memaddrs.len());
    assert_eq!(Some(&0), engine.module.borrow().memaddrs.get(0));

    assert_eq!(1, engine.store.memory.len());
    assert_eq!(10 * 1024 * 64, engine.store.memory[0].data.len());
    assert!(engine.store.memory[0].data.iter().all(|w| w == &0u8));
    assert_eq!(Some(20), engine.store.memory[0].max);
}

#[test]
fn test_allocation_globals() {
    let engine = allocation!(vec![Section::Global(GlobalSection {
        globals: vec![GlobalVariable {
            ty: GlobalType {
                value_type: ValueType::I32,
                mu: Mu::Const
            },
            init: vec![Instruction::Num(NumericInstructions::OP_I32_CONST(10))]
        }]
    })]);

    // Module instance has an entry in globaladdrs with 0
    // Store has a global instance

    assert_eq!(1, engine.module.borrow().globaladdrs.len());
    assert_eq!(Some(&0), engine.module.borrow().globaladdrs.get(0));

    assert_eq!(1, engine.store.globals.len());
    assert_eq!(
        Variable {
            mutable: false,
            val: Value::I32(10)
        },
        engine.store.globals[0]
    );
}

#[test]
fn test_allocation_exports() {
    let engine = allocation!(vec![
        Section::Memory(MemorySection {
            entries: vec![MemoryType {
                limits: Limits::Zero(10)
            }]
        }),
        Section::Export(ExportSection {
            entries: vec![ExportEntry {
                name: "memory".to_string(),
                kind: ExternalKindType::Memory { ty: 0 }
            }]
        })
    ]);

    // Module instance has an entry for exporsts

    assert_eq!(1, engine.module.borrow().exports.len());
    assert_eq!(
        ExportInstance {
            name: "memory".to_string(),
            value: ExternalKindType::Memory { ty: 0 }
        },
        engine.module.borrow().exports[0]
    );
}

#[test]
fn test_empty_wasm() {
    test_file_engine!("empty.wasm");
}

#[test]
fn test_sum_loop() {
    test_file_engine!("sum_loop.wasm");
}

#[test]
fn test_return_i32() {
    test_file_engine!("return_i32.wasm");
}

#[test]
fn test_return_i64() {
    test_file_engine!("return_i64.wasm");
}

#[test]
fn test_function_call() {
    test_file_engine!("function_call.wasm");
}

#[test]
fn test_arithmetic() {
    test_file_engine!("arithmetic.wasm");
}

#[test]
fn test_block_add_i32() {
    test_file_engine!("block_add_i32.wasm");
}

#[test]
fn test_loop_mult() {
    test_file_engine!("loop_mult.wasm");
}

#[test]
fn test_unreachable() {
    test_file_engine!("unreachable.wasm");
}

//#[test]
fn test_if_loop() {
    test_file_engine!("if_loop.wasm");
}

#[test]
fn test_logic() {
    test_file_engine!("logic.wasm");
}

#[test]
fn test_gcd() {
    test_file_engine!("gcd.wasm");
}

#[test]
fn test_run_add() {
    let engine = test_run_engine!("add.wasm", 0, vec![I32(1), I32(2)]);
    assert_eq!(
        Some(&StackContent::Value(I32(3))),
        engine.store.stack.last()
    )
}

#[test]
fn test_run_call() {
    /*
    (module
    (func $getAnswer (result i32) i32.const 42)
    (func (export "getAnswerPlus1") (result i32)
        call $getAnswer
        i32.const 1
        i32.add))
     */

    let engine = test_run_engine!("call.wasm", 0, vec![]);
    assert_eq!(
        Some(&StackContent::Value(I32(43))),
        engine.store.stack.last()
    )
}

fn test_run_gcd() {
    //env_logger::init();
    let engine = test_run_engine!("gcd.wasm", 2, vec![I32(50), I32(10)]);
    assert_eq!(
        Some(&StackContent::Value(I32(10))),
        engine.store.stack.last()
    )
}

//#[test]
fn test_run_incr_counter() {
    env_logger::init();
    let engine = test_run_engine!("incr_counter.wasm", 0, vec![]);
    assert_eq!(
        None,
        engine.store.stack.last()
    )
}

#[test]
fn test_run_gt() {
    let engine = test_run_engine!("gt.wasm", 0, vec![I32(10), I32(11)]);
    assert_eq!(
        Some(&StackContent::Value(I32(0))),
        engine.store.stack.last()
    )
}

#[test]
fn test_run_gt_2() {
    let engine = test_run_engine!("gt.wasm", 0, vec![I32(11), I32(10)]);
    assert_eq!(
        Some(&StackContent::Value(I32(1))),
        engine.store.stack.last()
    )
}

#[test]
fn test_run_sub() {
    let engine = test_run_engine!("sub.wasm", 0, vec![I32(10), I32(11)]);
    assert_eq!(
        Some(&StackContent::Value(I32(-1))),
        engine.store.stack.last()
    )
}

#[test]
fn test_run_sub2() {
    let engine = test_run_engine!("sub.wasm", 0, vec![I32(11), I32(10)]);
    assert_eq!(
        Some(&StackContent::Value(I32(1))),
        engine.store.stack.last()
    )
}

    /*
     (module
  (func (export "type-local-i32") (result i32) (local i32) (local.tee 0 (i32.const 0)))
  (func (export "type-local-i64") (result i64) (local i64) (local.tee 0 (i64.const 0)))
  (func (export "type-local-f32") (result f32) (local f32) (local.tee 0 (f32.const 0)))
  (func (export "type-local-f64") (result f64) (local f64) (local.tee 0 (f64.const 0)))

  (func (export "type-param-i32") (param i32) (result i32) (local.tee 0 (i32.const 10)))
  (func (export "type-param-i64") (param i64) (result i64) (local.tee 0 (i64.const 11)))
  (func (export "type-param-f32") (param f32) (result f32) (local.tee 0 (f32.const 11.1)))
  (func (export "type-param-f64") (param f64) (result f64) (local.tee 0 (f64.const 12.2)))

  (func (export "type-mixed") (param i64 f32 f64 i32 i32) (local f32 i64 i64 f64)
    (drop (i64.eqz (local.tee 0 (i64.const 0))))
    (drop (f32.neg (local.tee 1 (f32.const 0))))
    (drop (f64.neg (local.tee 2 (f64.const 0))))
    (drop (i32.eqz (local.tee 3 (i32.const 0))))
    (drop (i32.eqz (local.tee 4 (i32.const 0))))
    (drop (f32.neg (local.tee 5 (f32.const 0))))
    (drop (i64.eqz (local.tee 6 (i64.const 0))))
    (drop (i64.eqz (local.tee 7 (i64.const 0))))
    (drop (f64.neg (local.tee 8 (f64.const 0))))
  ))*/

//#[test]
fn test_run_local_tee() {
    let engine = test_run_engine!("local.tee.wasm", 0, vec![]);
    assert_eq!(
        Some(&StackContent::Value(I32(0))),
        engine.store.stack.last()
    )
}

//#[test]
fn test_run_local_tee_1() {
    let engine = test_run_engine!("local.tee.wasm", 1, vec![]);
    assert_eq!(
        Some(&StackContent::Value(I64(0))),
        engine.store.stack.last()
    )
}

//#[test]
fn test_run_local_tee_2() {
    let engine = test_run_engine!("local.tee.wasm", 2, vec![]);
    assert_eq!(
        Some(&StackContent::Value(F32(0.0))),
        engine.store.stack.last()
    )
}

//#[test]
fn test_run_local_tee_3() {
    let engine = test_run_engine!("local.tee.wasm", 3, vec![]);
    assert_eq!(
        Some(&StackContent::Value(F64(0.0))),
        engine.store.stack.last()
    )
}

//#[test]
fn test_run_local_tee_4() {
    let engine = test_run_engine!("local.tee.wasm", 4, vec![I32(2)]);
    assert_eq!(
        Some(&StackContent::Value(I32(10))),
        engine.store.stack.last()
    )
}

//#[test]
fn test_run_local_tee_5() {
    let engine = test_run_engine!("local.tee.wasm", 5, vec![I64(3)]);
    assert_eq!(
        Some(&StackContent::Value(I64(11))),
        engine.store.stack.last()
    )
}

//#[test]
fn test_run_local_tee_6() {
    let engine = test_run_engine!("local.tee.wasm", 6, vec![F32(4.4)]);
    assert_eq!(
        Some(&StackContent::Value(F32(11.1))),
        engine.store.stack.last()
    )
}

//#[test]
fn test_run_local_tee_7() {
    let engine = test_run_engine!("local.tee.wasm", 7, vec![F64(5.5)]);
    assert_eq!(
        Some(&StackContent::Value(F64(12.2))),
        engine.store.stack.last()
    )
}

//#[test]
fn test_run_as_loop_first_br_if_1() {
    /*
      (func (export "as-loop-first") (param i32) (result i32)
        (block (loop (br_if 1 (local.get 0)) (return (i32.const 2)))) (i32.const 3)
      )
    */
    let engine = test_run_engine!("as_loop_br_if.wasm", 0, vec![I32(0)]);
    assert_eq!(
        Some(&StackContent::Value(I32(2))),
        engine.store.stack.last()
    )
}


//#[test]
fn test_run_as_loop_first_br_if_2() {
    /*
      (func (export "as-loop-first") (param i32) (result i32)
        (block (loop (br_if 1 (local.get 0)) (return (i32.const 2)))) (i32.const 3)
      )
    */
//    env_logger::init();
    let engine = test_run_engine!("as_loop_br_if.wasm", 0, vec![I32(1)]);
    assert_eq!(
        Some(&StackContent::Value(I32(3))),
        engine.store.stack.last()
    )
}

/*
(func (export "br_if0") (result i32)
    (local $i i32)
    (local.set $i (i32.const 0))
    (block $outer (result i32)
      (block $inner
        (br_if $inner (i32.const 0))
        (local.set $i (i32.or (local.get $i) (i32.const 0x1)))
        (br_if $inner (i32.const 1))
        (local.set $i (i32.or (local.get $i) (i32.const 0x2)))
      )
      (drop (br_if $outer
        (block (result i32)
          (local.set $i (i32.or (local.get $i) (i32.const 0x4)))
          (local.get $i)
        )
        (i32.const 0)
      ))
      (local.set $i (i32.or (local.get $i) (i32.const 0x8)))
      (drop (br_if $outer
        (block (result i32)
          (local.set $i (i32.or (local.get $i) (i32.const 0x10)))
          (local.get $i)
        )
        (i32.const 1)
      ))
      (local.set $i (i32.or (local.get $i) (i32.const 0x20))) (local.get $i)
    )
  )

  (func (export "br_if1") (result i32)
    (block $l0 (result i32)
      (drop
        (br_if $l0
          (block $l1 (result i32) (br $l1 (i32.const 1)))
          (i32.const 1)
        )
      )
      (i32.const 0)
    )
  )

  (func (export "br_if2") (result i32)
    (block $l0 (result i32)
      (if (i32.const 1)
        (then
          (drop
            (br_if $l0
              (block $l1 (result i32) (br $l1 (i32.const 1)))
              (i32.const 1)
            )
          )
        )
      )
      (i32.const 0)
    )
  )

  (func (export "br_if3") (result i32)
    (local $i1 i32)
    (drop
      (i32.add
        (block $l0 (result i32)
          (drop (br_if $l0
            (block (result i32) (local.set $i1 (i32.const 1)) (local.get $i1))
            (block (result i32) (local.set $i1 (i32.const 2)) (local.get $i1))
          ))
          (i32.const 0)
        )
        (i32.const 0)
      )
    )
    (local.get $i1)
  )

  (func (export "br") (result i32)
    (block $l0 (result i32)
      (if (i32.const 1)
        (then (br $l0 (block $l1 (result i32) (br $l1 (i32.const 1)))))
        (else (block (drop (block $l1 (result i32) (br $l1 (i32.const 1))))))
      )
      (i32.const 1)
    )
  ))*/

//#[test]
fn test_run_br_if0() {
    //env_logger::init();
    let engine = test_run_engine!("labels.wasm", 0, vec![]);
    assert_eq!(
        Some(&StackContent::Value(I32(0x1d))),
        engine.store.stack.last()
    )
}

//#[test]
fn test_run_br_if1() {
    //env_logger::init();
    let engine = test_run_engine!("labels.wasm", 1, vec![]);
    assert_eq!(
        Some(&StackContent::Value(I32(0x1))),
        engine.store.stack.last()
    )
}

//#[test]
fn test_run_br_if2() {
    //env_logger::init();
    let engine = test_run_engine!("labels.wasm", 2, vec![]);
    assert_eq!(
        Some(&StackContent::Value(I32(1))),
        engine.store.stack.last()
    )
}

fn test_run_br_if3() {
    //env_logger::init();
    let engine = test_run_engine!("labels.wasm", 3, vec![]);
    assert_eq!(
        Some(&StackContent::Value(I32(1))),
        engine.store.stack.last()
    )
}

fn test_run_br_if4() {
    //env_logger::init();
    let engine = test_run_engine!("labels.wasm", 4, vec![]);
    assert_eq!(
        Some(&StackContent::Value(I32(2))),
        engine.store.stack.last()
    )
}

/*
(func (export "as-loop-mid") (param i32) (result i32)
    (block (loop (call $dummy) (br_if 1 (local.get 0)) (return (i32.const 2))))
    (i32.const 4)
  )
*/

//#[test]
fn test_run_as_loop_mid_br1() {
    //env_logger::init();
    let engine = test_run_engine!("as_loop_mid_br_if.wasm", 0, vec![I32(0)]);
    assert_eq!(
        Some(&StackContent::Value(I32(2))),
        engine.store.stack.last()
    )
}

//#[test]
fn test_run_as_loop_mid_br2() {
    //env_logger::init();
    let engine = test_run_engine!("as_loop_mid_br_if.wasm", 0, vec![I32(1)]);
    assert_eq!(
        Some(&StackContent::Value(I32(4))),
        engine.store.stack.last()
    )
}

/*
  (func (export "as-loop-last") (param i32)
    (loop (call $dummy) (br_if 1 (local.get 0)))
  )
*/

//#[test]
/*
fn test_run_as_loop_last_br_if1() {
    env_logger::init();
    let engine = test_run_engine!("as_loop_last_br_if.wasm", 0, vec![I32(5)]);
    assert_eq!(
        None,
        engine.store.stack.last()
    )
}*/

//#[test]
fn test_run_as_loop_last_br_if2() {
    //env_logger::init();
    let engine = test_run_engine!("as_loop_last_br_if.wasm", 0, vec![I32(1)]);
    assert_eq!(
        None,
        engine.store.stack.last()
    )
}


/*
#[test]
fn test_run_call_indirect() {
    /*
     (module
      (table 2 anyfunc)
      (func $f1 (result i32)
        i32.const 42)
      (func $f2 (result i32)
        i32.const 13)
      (elem (i32.const 0) $f1 $f2)
      (type $return_i32 (func (result i32)))
      (func (export "callByIndex") (param $i i32) (result i32)
        local.get $i
        call_indirect (type $return_i32))
    )*/

    env_logger::init();
    let engine = test_run_engine!("wasm-table.wasm", 0, vec![I32(1)]);
    assert_eq!(
        Some(&StackContent::Value(I32(43))),
        engine.store.stack.last()
    )
}
*/
