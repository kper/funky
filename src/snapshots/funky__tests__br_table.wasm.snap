---
source: src/tests.rs
expression: "format!(\"{:#?}\", engine)"
---
Engine {
    module: RefCell {
        value: ModuleInstance {
            start: 0,
            code: [
                FunctionBody {
                    locals: [],
                    code: [
                        Ctrl(
                            OP_BLOCK(
                                Empty,
                                [
                                    Ctrl(
                                        OP_BLOCK(
                                            Empty,
                                            [
                                                Ctrl(
                                                    OP_BLOCK(
                                                        Empty,
                                                        [
                                                            Ctrl(
                                                                OP_BLOCK(
                                                                    Empty,
                                                                    [
                                                                        Ctrl(
                                                                            OP_BLOCK(
                                                                                Empty,
                                                                                [
                                                                                    Var(
                                                                                        OP_LOCAL_GET(
                                                                                            0,
                                                                                        ),
                                                                                    ),
                                                                                    Ctrl(
                                                                                        OP_BR_TABLE(
                                                                                            [
                                                                                                3,
                                                                                                2,
                                                                                                1,
                                                                                                0,
                                                                                            ],
                                                                                            4,
                                                                                        ),
                                                                                    ),
                                                                                    Num(
                                                                                        OP_I32_CONST(
                                                                                            99,
                                                                                        ),
                                                                                    ),
                                                                                    Ctrl(
                                                                                        OP_RETURN,
                                                                                    ),
                                                                                ],
                                                                            ),
                                                                        ),
                                                                        Num(
                                                                            OP_I32_CONST(
                                                                                100,
                                                                            ),
                                                                        ),
                                                                        Ctrl(
                                                                            OP_RETURN,
                                                                        ),
                                                                    ],
                                                                ),
                                                            ),
                                                            Num(
                                                                OP_I32_CONST(
                                                                    101,
                                                                ),
                                                            ),
                                                            Ctrl(
                                                                OP_RETURN,
                                                            ),
                                                        ],
                                                    ),
                                                ),
                                                Num(
                                                    OP_I32_CONST(
                                                        102,
                                                    ),
                                                ),
                                                Ctrl(
                                                    OP_RETURN,
                                                ),
                                            ],
                                        ),
                                    ),
                                    Num(
                                        OP_I32_CONST(
                                            103,
                                        ),
                                    ),
                                    Ctrl(
                                        OP_RETURN,
                                    ),
                                ],
                            ),
                        ),
                        Num(
                            OP_I32_CONST(
                                104,
                            ),
                        ),
                    ],
                },
            ],
            fn_types: [
                FunctionSignature {
                    param_types: [
                        I32,
                    ],
                    return_types: [
                        I32,
                    ],
                },
            ],
            funcaddrs: [
                0,
            ],
            tableaddrs: [],
            memaddrs: [],
            globaladdrs: [],
            exports: [
                ExportInstance {
                    name: "multiple",
                    value: Function {
                        ty: 0,
                    },
                },
                ExportInstance {
                    name: "br_table",
                    value: Function {
                        ty: 0,
                    },
                },
            ],
        },
    },
    started: false,
    store: Store {
        funcs: [
            FuncInstance {
                ty: FunctionSignature {
                    param_types: [
                        I32,
                    ],
                    return_types: [
                        I32,
                    ],
                },
                module: (Weak),
                code: FunctionBody {
                    locals: [],
                    code: [
                        Ctrl(
                            OP_BLOCK(
                                Empty,
                                [
                                    Ctrl(
                                        OP_BLOCK(
                                            Empty,
                                            [
                                                Ctrl(
                                                    OP_BLOCK(
                                                        Empty,
                                                        [
                                                            Ctrl(
                                                                OP_BLOCK(
                                                                    Empty,
                                                                    [
                                                                        Ctrl(
                                                                            OP_BLOCK(
                                                                                Empty,
                                                                                [
                                                                                    Var(
                                                                                        OP_LOCAL_GET(
                                                                                            0,
                                                                                        ),
                                                                                    ),
                                                                                    Ctrl(
                                                                                        OP_BR_TABLE(
                                                                                            [
                                                                                                3,
                                                                                                2,
                                                                                                1,
                                                                                                0,
                                                                                            ],
                                                                                            4,
                                                                                        ),
                                                                                    ),
                                                                                    Num(
                                                                                        OP_I32_CONST(
                                                                                            99,
                                                                                        ),
                                                                                    ),
                                                                                    Ctrl(
                                                                                        OP_RETURN,
                                                                                    ),
                                                                                ],
                                                                            ),
                                                                        ),
                                                                        Num(
                                                                            OP_I32_CONST(
                                                                                100,
                                                                            ),
                                                                        ),
                                                                        Ctrl(
                                                                            OP_RETURN,
                                                                        ),
                                                                    ],
                                                                ),
                                                            ),
                                                            Num(
                                                                OP_I32_CONST(
                                                                    101,
                                                                ),
                                                            ),
                                                            Ctrl(
                                                                OP_RETURN,
                                                            ),
                                                        ],
                                                    ),
                                                ),
                                                Num(
                                                    OP_I32_CONST(
                                                        102,
                                                    ),
                                                ),
                                                Ctrl(
                                                    OP_RETURN,
                                                ),
                                            ],
                                        ),
                                    ),
                                    Num(
                                        OP_I32_CONST(
                                            103,
                                        ),
                                    ),
                                    Ctrl(
                                        OP_RETURN,
                                    ),
                                ],
                            ),
                        ),
                        Num(
                            OP_I32_CONST(
                                104,
                            ),
                        ),
                    ],
                },
            },
        ],
        tables: [],
        memory: [],
        stack: [],
        globals: [],
    },
}
